### DISCUSSION TITLE
Use sqlite vector add-on (for embeddings)?

### DISCUSSION THREAD

**Discussion by xrd** [pocketbase/pocketbase#4632] [Q&A] [answered]
Would it be possible to use a vector addition to sqlite? For example:

https://github.com/asg017/sqlite-vss

I imagine there would be serious limitations in the JS/golang libraries. But, I see examples of using virtual tables (meaning you could make it fit inside the API signature with a little extra work), and wonder if this is still worthwhile despite the limitations?

Basically, I want to use vector embeddings inside sqlite. I'm open to another plugin if there is a better fit.

**Comment by ganigeorgiev** [comment, marked as answer]
Loading runtime SQLite extension or entire custom SQLite build is not possible at the moment but it will be supported with the [refactoring](https://github.com/pocketbase/pocketbase/discussions/4355) at least when the CGO driver is used (_for the pure Go SQLite driver the extension will have to be transpiled first with [ccgo](https://gitlab.com/cznic/ccgo), if possible at all_).

**Reply by xrd** [reply]
OK, great! Is there a preview release I can experiment with now to align myself with this new way of doing things? It looks very exciting!

**Reply by xrd** [reply]
Also, if I ran a sqlite binary (that was not the pocketbase golang binary) would it cause problems if I loaded the extension there? 

Even if the pocketbase binary cannot support exposing vector tables, I could see value in keeping all my data inside the same sqlite database file.

**Reply by ganigeorgiev** [reply]
No, I'm still experimenting and I haven't published anything yet because most of the changes are intertwined and depends on one another or require a shared basic functionality first to be implemented in attempt to produce simpler and cleaner public interfaces (_for example the hooks are refactored and everything "hook"-able in PocketBase will use the same signature including the new router handlers and middlewares_), so it will take some time before I can show the results of the refactoring.

**Reply by ganigeorgiev** [reply]
> Also, if I ran a sqlite binary (that was not the pocketbase golang binary) would it cause problems if I loaded the extension there?

I'm not sure that I understand what do you mean by _sqlite binary_ in this context but you can initialize and open connections to other databases when using Go just fine but they'll not show up in the PocketBase Admin UI or be "interactable" with the builtin web APIs.

**Reply by xrd** [reply]
I mean if I opened a shell and installed the sqlite3 binary from ubuntu and then ran `sqlite3 data.db` and then 

```
.load ./vector0
.load ./vss0

select vss_version(); -- 'v0.0.1'
```

Would that "break" the data.db file so pocketbase could not use it normally?

**Reply by ganigeorgiev** [reply]
Hm, I'm not sure it will have to be tested.

Probably it will also depend on what the extension does and how it operates as some extensions can "stick" after the original process and connection has been closed. From the [SQLite docs](https://www.sqlite.org/loadext.html#persistent_loadable_extensions):

> The default behavior for a loadable extension is that it is unloaded from process memory when the database connection that originally invoked [sqlite3_load_extension()](https://www.sqlite.org/c3ref/load_extension.html) closes. (In other words, the xDlClose method of the [sqlite3_vfs](https://www.sqlite.org/c3ref/vfs.html) object is called for all extensions when a database connection closes.) However, if the initialization procedure returns [SQLITE_OK_LOAD_PERMANENTLY](https://www.sqlite.org/rescode.html#ok_load_permanently) instead of SQLITE_OK, then the extension will not be unloaded (xDlClose will not be invoked) and the extension will remain in process memory indefinitely.

_edit: On second read, the above should matter only for the sqlite3 process so it shouldn't affect the PocketBase one._

**Reply by xrd** [reply]
I'll report back on what I find. Thanks so much for your comments!

**Reply by xrd** [reply]
Just an FYI. I tried to use the vss extension on a regular sqlite database (without using pocketbase) at all. And, I found that there were many weird bugs (like https://github.com/asg017/sqlite-vss/issues/91) that made me feel like this extension is not ready for primetime. Instead, I've been happy with chromadb (https://docs.trychroma.com/). ChromaDB, incidentally, uses sqlite under the hood if you persist the data, and I'll see whether I can coerce pocketbase to see that data at another time.

**Reply by mukhtharcm** [reply]
so you're running chroma alongside pocketbase?
do you have a guide or something?

**Reply by rodydavis** [reply]
I made a PR for loadable extensions:
https://github.com/pocketbase/pocketbase/pull/5397

I also got sqlite-vec working with PocketBase and creating an example to share.

You can also load the extension before you create the PocketBase app and it works just fine for me with CGO:

`sqlite_vec.Auto()`

**Reply by mukhtharcm** [reply]
That's awesome!

I see that the pr is closed
but i'd like to see that example repo :)

**Reply by rodydavis** [reply]
Here is the example, it also doesn't need the PR!

<details>
<summary>Code Example</summary>

```go
package plugins

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"strconv"

	sqlite_vec "github.com/asg017/sqlite-vec-go-bindings/cgo"
	"github.com/google/generative-ai-go/genai"
	"github.com/labstack/echo/v5"
	"github.com/pocketbase/dbx"
	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/apis"
	"github.com/pocketbase/pocketbase/core"
	"github.com/pocketbase/pocketbase/models"
	"github.com/pocketbase/pocketbase/models/schema"
	"github.com/pocketbase/pocketbase/tools/types"
	"google.golang.org/api/option"
)

func VectorStore(app *pocketbase.PocketBase, collections ...string) error {
	sqlite_vec.Auto()

	client, err := createGoogleAiClient()
	if err != nil {
		return err
	}

	app.OnAfterBootstrap().Add(func(e *core.BootstrapEvent) error {
		for _, target := range collections {
			collection, _ := app.Dao().FindCollectionByNameOrId(target)
			if collection == nil {
				err := createVectorCollection(app, target)
				if err != nil {
					return err
				}
			}
		}
		return nil
	})

	app.OnModelAfterCreate().Add(func(e *core.ModelEvent) error {
		tbl := e.Model.TableName()
		for _, target := range collections {
			if tbl == target {
				err := modelModify(app, target, client, e)
				if err != nil {
					return err
				}
			}
		}
		return nil
	})
	app.OnModelAfterUpdate().Add(func(e *core.ModelEvent) error {
		tbl := e.Model.TableName()
		for _, target := range collections {
			if tbl == target {
				err := modelModify(app, target, client, e)
				if err != nil {
					return err
				}
			}
		}
		return nil
	})
	app.OnModelAfterDelete().Add(func(e *core.ModelEvent) error {
		tbl := e.Model.TableName()
		for _, target := range collections {
			if tbl == target {
				err := modelDelete(app, target, e)
				if err != nil {
					return err
				}
			}
		}
		return nil
	})
	app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
		group := e.Router.Group("/api/collections/:collectionIdOrName/records", apis.ActivityLogger(app))
		group.GET("/vector-search", func(c echo.Context) error {
			target := c.PathParam("collectionIdOrName")
			if _, err := app.Dao().FindCollectionByNameOrId(target); err != nil {
				app.Logger().Error(fmt.Sprint(err))
				return err
			}

			title := c.QueryParam("title")
			content := c.QueryParam("content")
			k := c.QueryParam("k")
			kNum := 5
			if k != "" {
				val, err := strconv.Atoi(k)
				if err == nil {
					kNum = val
				}
			}

			if content == "" {
				return c.NoContent(204)
			}

			vector, err := googleAiEmbedContent(client, genai.TaskTypeRetrievalQuery, title, genai.Text(content))
			if err != nil {
				return err
			}
			jsonVec, err := json.Marshal(vector)
			if err != nil {
				return err
			}

			stmt := "SELECT v.id, distance, v.content, v.created, v.updated "
			stmt += "FROM " + target + "_embeddings "
			stmt += "LEFT JOIN " + target + " v ON v.vector_id = " + target + "_embeddings.id "
			stmt += "WHERE embedding MATCH {:embedding} "
			stmt += "AND k = {:k};"

			results := []dbx.NullStringMap{}
			err = app.Dao().DB().
				NewQuery(stmt).
				Bind(dbx.Params{
					"embedding": string(jsonVec),
					"k":         kNum,
				}).
				All(&results)
			if err != nil {
				app.Logger().Error(fmt.Sprint(err))
				return err
			}
			app.Logger().Info(fmt.Sprint(results))

			c.Response().Header().Set(echo.HeaderContentType, "application/json")
			items := []map[string]any{}
			for _, result := range results {
				m := make(map[string]interface{})
				for key := range result {
					val := result[key]
					value, err := val.Value()
					if err != nil || !val.Valid {
						m[key] = nil
					} else {
						m[key] = value
					}
				}
				items = append(items, m)
			}

			// TODO: Paging result
			return c.JSON(200, items)

		})
		return nil
	})
	return nil
}

func modelDelete(app *pocketbase.PocketBase, target string, e *core.ModelEvent) error {
	_, err := e.Dao.FindRecordById(e.Model.TableName(), e.Model.GetId())
	if err != nil {
		return err
	}
	deleteEmbeddingsForRecord(app, target, e)
	return nil
}

func modelModify(app *pocketbase.PocketBase, target string, client *genai.Client, e *core.ModelEvent) error {
	record, err := e.Dao.FindRecordById(e.Model.TableName(), e.Model.GetId())
	if err != nil {
		return err
	}
	title := record.GetString("title")
	content := record.GetString("content")

	result, err := googleAiEmbedContent(client, genai.TaskTypeRetrievalDocument, title, genai.Text(content))
	if err != nil {
		return err
	}

	vector := ""
	jsonVec, err := json.Marshal(result)
	if err != nil {
		vector = "[]"
	} else {
		vector = string(jsonVec)
	}

	deleteEmbeddingsForRecord(app, target, e)

	{
		stmt := "INSERT INTO " + target + "_embeddings (embedding) "
		stmt += "VALUES ({:embedding});"
		res, err := app.DB().NewQuery(stmt).Bind(dbx.Params{
			"embedding": vector,
		}).Execute()
		if err != nil {
			return nil
		}
		vectorId, err := res.LastInsertId()
		if err != nil {
			return err
		}
		record.Set("vector_id", vectorId)
	}

	if err := app.Dao().WithoutHooks().SaveRecord(record); err != nil {
		return err
	}
	return nil
}

func deleteEmbeddingsForRecord(app *pocketbase.PocketBase, target string, e *core.ModelEvent) error {
	record, err := e.Dao.FindRecordById(e.Model.TableName(), e.Model.GetId())
	if err != nil {
		return err
	}

	type Meta struct {
		Id string `db:"id" json:"id"`
	}
	vectorId := record.GetInt("vector_id")
	items := []*Meta{}
	stmt := "SELECT id FROM " + target + "_embeddings "
	stmt += "WHERE id = {:id};"
	err = app.DB().NewQuery(stmt).Bind(dbx.Params{
		"id": vectorId,
	}).All(&items)
	if err != nil {
		return nil
	}

	stmt = "DELETE FROM " + target + "_embeddings "
	stmt += "WHERE id = {:id}"

	for _, item := range items {
		_, err = app.DB().NewQuery(stmt).Bind(dbx.Params{
			"id": item.Id,
		}).Execute()
		if err != nil {
			return nil
		}
	}

	return nil
}

func createVectorCollection(app *pocketbase.PocketBase, target string, extraFields ...*schema.SchemaField) error {
	fields := []*schema.SchemaField{
		{
			Name: "title",
			Type: schema.FieldTypeText,
		},
		{
			Name:     "content",
			Required: true,
			Type:     schema.FieldTypeText,
		},
		{
			Name: "vector_id",
			Type: schema.FieldTypeNumber,
		},
	}
	fields = append(fields, extraFields...)
	collection := &models.Collection{
		Name:   target,
		Type:   models.CollectionTypeBase,
		Schema: schema.NewSchema(fields...),
		Indexes: types.JsonArray[string]{
			"CREATE UNIQUE INDEX idx_" + target + " ON " + target + " (title, content, vector_id);",
		},
	}

	if err := app.Dao().SaveCollection(collection); err != nil {
		return err
	}

	stmt := "CREATE VIRTUAL TABLE IF NOT EXISTS " + target + "_embeddings using vec0( "
	stmt += "	id INTEGER PRIMARY KEY AUTOINCREMENT, "
	stmt += "	embedding float[768] "
	stmt += ");"
	_, err := app.DB().NewQuery(stmt).Execute()
	if err != nil {
		return nil
	}

	return nil
}

func createGoogleAiClient() (*genai.Client, error) {
	var apiKey string = os.Getenv("GOOGLE_AI_API_KEY")
	ctx := context.Background()
	client, err := genai.NewClient(ctx, option.WithAPIKey(apiKey))
	if err != nil {
		return nil, err
	}
	return client, nil
}

func googleAiEmbedContent(client *genai.Client, taskType genai.TaskType, title string, parts ...genai.Part) ([]float32, error) {
	ctx := context.Background()
	model := googleAiEmbeddingModel(client)
	model.TaskType = taskType
	res, err := model.EmbedContentWithTitle(ctx, title, parts...)
	if err != nil {
		return nil, err
	}
	return res.Embedding.Values, nil
}

func googleAiEmbeddingModel(client *genai.Client) *genai.EmbeddingModel {
	em := client.EmbeddingModel("text-embedding-004")
	return em
}

```

</details>

https://gist.github.com/rodydavis/e80a5919ca531793ddbf86fecf8538e3

Usage:

```go
func main() {
	app := pocketbase.New()
	
	err := plugins.VectorStore(app, "vectors")
	if err != nil {
		log.Fatal(err)
	}
	
	if err := app.Start(); err != nil {
		log.Fatal(err)
	}
}
```
